framework/helpers/docker.py

```diff
@@ -0,0 +1,295 @@
+import datetime
+import math
+import pathlib
+import queue
+import threading
+
+import absl
+import grpc
+import mako.template
+
+import docker
+from protos.grpc.testing import messages_pb2
+from protos.grpc.testing import test_pb2_grpc
+from protos.grpc.testing.xdsconfig.control_pb2 import StopOnRequestRequest
+from protos.grpc.testing.xdsconfig.control_pb2 import UpsertResourcesRequest
+from protos.grpc.testing.xdsconfig.service_pb2_grpc import (
+    XdsConfigControlServiceStub,
+)
+
+
+class Bootstrap:
+    def __init__(self, base: str, ports: list[int], host_name: str):
+        self.base = pathlib.PosixPath(base)
+        self.host_name = host_name
+        self.ports = ports
+        self.mount_dir: pathlib.Path = None
+        self.MakeWorkingDir(self.base)
+        # Use Mako
+        template = mako.template.Template(filename="templates/bootstrap.json")
+        file = template.render(
+            servers=[f"{self.host_name}:{port}" for port in self.ports]
+        )
+        destination = self.get_mount_dir() / "bootstrap.json"
+        with open(destination, "w", encoding="utf-8") as f:
+            f.write(file)
+            absl.logging.debug("Generated bootstrap file at %s", destination)
+
+    def MakeWorkingDir(self, base: str):
+        # Date time to string
+        run_id = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
+        self.mount_dir = base / f"testrun_{run_id}"
+        if not self.mount_dir.exists():
+            absl.logging.debug("Creating %s", self.mount_dir)
+            self.get_mount_dir().mkdir(parents=True)
+
+    def get_mount_dir(self):
+        if self.mount_dir is None:
+            raise RuntimeError("Working dir was not created yet")
+        return self.mount_dir.absolute()
+
+    def xds_config_server_port(self, n: int):
+        return self.ports[n]
+
+
+class ChildProcessEvent:
+    def __init__(self, source: str, data: str):
+        self.source = source
+        self.data = data
+
+    def __str__(self) -> str:
+        return f"{self.data}, {self.source}"
+
+    def __repr__(self) -> str:
+        return f"data={self.data}, source={self.source}"
+
+
+class ProcessManager:
+    def __init__(
+        self,
+        bootstrap: Bootstrap,
+        nodeId: str,
+        verbosity="info",
+    ):
+        self.dockerClient = docker.DockerClient.from_env()
```

Comment:
Tip on fixing pylint no-member issues: figure out in what file the class is defined, and import it. In this case, you want

```
import docker.client
```

And then refer refer to `DockerClient` as `docker.client.DockerClient`.